File: defender/ent/schema/report.go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"github.com/google/uuid"
)

type Report struct {
	ent.Schema
}

func (Report) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).Default(uuid.New),
		field.JSON("metas", map[string]any{}).Optional(),
		field.JSON("request_headers", []map[string]string{}).Optional(),
		field.JSON("request_body", map[string]any{}).Optional(),
		field.JSON("response_headers", []map[string]string{}).Optional(),
		field.JSON("response_body", map[string]any{}).Optional(),
		field.JSON("rule_details", map[string]any{}).Optional(),
		field.UUID("triggered_by", uuid.UUID{}).Optional().Nillable(),
		field.UUID("created_by", uuid.UUID{}).Optional().Nillable(),
		field.Time("created_at").Default(time.Now).Immutable(),
		field.Time("updated_at").Default(time.Now).UpdateDefault(time.Now),
	}
}

func (Report) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("triggered_by_action", Action.Type).
			Ref("reports").
			Field("triggered_by").
			Unique(),
		edge.From("created_by_defender", Defender.Type).
			Ref("reports").
			Field("created_by").
			Unique(),
	}
}

File: manager/app/Models/Report.php
<?php

namespace App\Models;

use App\Observers\ReportObserver;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Attributes\ObservedBy;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;

#[Fillable(['metas', 'request_headers', 'request_body', 'response_headers', 'response_body', 'rule_details', 'triggered_by', 'created_by', 'is_reviewed'])]
#[ObservedBy(ReportObserver::class)]
class Report extends Model
{
    use HasUuids;

    protected function casts(): array
    {
        return [
            'id' => 'string',
            'metas' => 'array',
            'request_headers' => 'array',
            'request_body' => 'array',
            'response_headers' => 'array',
            'response_body' => 'array',
            'rule_details' => 'array',
            'triggered_by' => 'string',
            'created_by' => 'string',
            'is_reviewed' => 'boolean',
            'created_at' => 'datetime',
            'updated_at' => 'datetime',
        ];
    }

    public function triggeredBy()
    {
        return $this->belongsTo(Action::class, 'triggered_by');
    }

    public function createdBy()
    {
        return $this->belongsTo(Defender::class, 'created_by');
    }
}
